mesonbuild/modules/unstable_icestorm.py

```diff
@@ -12,78 +12,115 @@
+                args: T.Tuple[str, T.List[T.Union[mesonlib.FileOrString, build.GeneratedTypes]]],
+                kwargs: ProjectKwargs) -> ModuleReturnValue:
+        if not self.tools:
             self.detect_binaries(state)
-        if not args:
-            raise mesonlib.MesonException('Project requires at least one argument, which is the project name.')
-        proj_name = args[0]
-        arg_sources = args[1:]
-        if not isinstance(proj_name, str):
-            raise mesonlib.MesonException('Argument must be a string.')
-        kwarg_sources = kwargs.get('sources', [])
-        if not isinstance(kwarg_sources, list):
-            kwarg_sources = [kwarg_sources]
-        all_sources = self.interpreter.source_strings_to_files(flatten(arg_sources + kwarg_sources))
-        if 'constraint_file' not in kwargs:
-            raise mesonlib.MesonException('Constraint file not specified.')
-
-        constraint_file = self.interpreter.source_strings_to_files(kwargs['constraint_file'])
-        if len(constraint_file) != 1:
-            raise mesonlib.MesonException('Constraint file must contain one and only one entry.')
-        blif_name = proj_name + '_blif'
-        blif_fname = proj_name + '.blif'
-        asc_name = proj_name + '_asc'
-        asc_fname = proj_name + '.asc'
-        bin_name = proj_name + '_bin'
-        bin_fname = proj_name + '.bin'
-        time_name = proj_name + '-time'
-        upload_name = proj_name + '-upload'
-
-        blif_target = self.interpreter.func_custom_target(None, [blif_name], {
-            'input': all_sources,
-            'output': blif_fname,
-            'command': [self.yosys_bin, '-q', '-p', 'synth_ice40 -blif @OUTPUT@', '@INPUT@']})
-
-        asc_target = self.interpreter.func_custom_target(None, [asc_name], {
-            'input': blif_target,
-            'output': asc_fname,
-            'command': [self.arachne_bin, '-q', '-d', '1k', '-p', constraint_file, '@INPUT@', '-o', '@OUTPUT@']})
-
-        bin_target = self.interpreter.func_custom_target(None, [bin_name], {
-            'input': asc_target,
-            'output': bin_fname,
-            'command': [self.icepack_bin, '@INPUT@', '@OUTPUT@'],
-            'build_by_default': True})
-
-        self.interpreter.func_run_target(None, [upload_name], {
-            'command': [self.iceprog_bin, bin_target]})
-
-        self.interpreter.func_run_target(None, [time_name], {
-            'command': [self.icetime_bin, bin_target]})
-
-def initialize(*args, **kwargs):
-    return IceStormModule(*args, **kwargs)
+        proj_name, arg_sources = args
+        all_sources = self.interpreter.source_strings_to_files(
+            list(itertools.chain(arg_sources, kwargs['sources'])))
+
+        blif_target = build.CustomTarget(
+            f'{proj_name}_blif',
+            state.subdir,
+            state.subproject,
+            [self.tools['yosys'], '-q', '-p', 'synth_ice40 -blif @OUTPUT@', '@INPUT@'],
+            all_sources,
```

Comment:
mypy thinks that all_sources might contain StructuredSources here.

Reply:
Yeah, I was unmotivated to tackle it

Reply:
I’ll deal with it Monday

Reply:
ping?

Reply:
Ping?